File: sdk/js/src/identity.ts
/**
 * Local agent identity storage and Verifier registration (Node.js).
 * Browser runtimes must inject TAP_PRIVATE_KEY + TAP_KID or pass identityDir
 * to a Node-compatible storage backend.
 */
import { createHash } from "node:crypto";
import { chmod, mkdir, readFile, writeFile } from "node:fs/promises";
import { homedir } from "node:os";
import path from "node:path";

import { generateSigner, newId, publicJwk, type Jwk } from "./tap.js";

export interface AgentIdentity {
  agentId: string;
  kid: string;
  seedHex: string;
}

const cache = new Map<string, AgentIdentity>();

export function defaultIdentityDir(): string {
  return process.env.TAP_IDENTITY_DIR?.trim() || path.join(homedir(), ".tap", "identities");
}

function namespace(apiKey: string | undefined, endpoint: string): string {
  return createHash("sha256").update(`${endpoint}\0${apiKey ?? ""}`).digest("hex").slice(0, 16);
}

function identityFile(identityDir: string, ns: string, agentId: string): string {
  const safe = agentId.replace(/\//g, "_").replace(/\\/g, "_");
  return path.join(identityDir, ns, `${safe}.json`);
}

function fromEnv(agentId: string): AgentIdentity | null {
  const envAgent = process.env.TAP_AGENT_ID ?? process.env.TAP_AGENT_ID;
  const seed = process.env.TAP_PRIVATE_KEY ?? process.env.TAP_PRIVATE_KEY_HEX;
  const kid = process.env.TAP_KID;
  if (!seed || !kid) return null;
  if (envAgent && envAgent !== agentId) return null;
  return { agentId, kid, seedHex: seed.replace(/^0x/, "") };
}

async function loadFile(filePath: string): Promise<AgentIdentity | null> {
  try {
    const data = JSON.parse(await readFile(filePath, "utf8")) as {
      agent_id: string;
      kid: string;
      seed_hex: string;
    };
    return { agentId: data.agent_id, kid: data.kid, seedHex: data.seed_hex };
  } catch {
    return null;
  }
}

async function saveFile(filePath: string, identity: AgentIdentity): Promise<void> {
  await mkdir(path.dirname(filePath), { recursive: true });
  await writeFile(
    filePath,
    `${JSON.stringify(
      { agent_id: identity.agentId, kid: identity.kid, seed_hex: identity.seedHex },
      null,
      2,
    )}\n`,
    { mode: 0o600 },
  );
  try {
    await chmod(filePath, 0o600);
  } catch {
    /* windows */
  }
}

export type RegisterFn = (args: {
  endpoint: string;
  apiKey?: string;
  agentId: string;
  publicJwk: Jwk;
}) => Promise<string>;

export async function registerWithVerifier(args: {
  endpoint: string;
  apiKey?: string;
  agentId: string;
  publicJwk: Jwk;
  registerFn?: RegisterFn;
}): Promise<string> {
  if (args.registerFn) {
    return args.registerFn({
      endpoint: args.endpoint,
      apiKey: args.apiKey,
      agentId: args.agentId,
      publicJwk: args.publicJwk,
    });
  }
  const headers: Record<string, string> = { "Content-Type": "application/json" };
  if (args.apiKey) headers["X-API-Key"] = args.apiKey;
  const res = await fetch(`${args.endpoint.replace(/\/$/, "")}/v1/agents/register`, {
    method: "POST",
    headers,
    body: JSON.stringify({ agent_id: args.agentId, public_jwk: args.publicJwk }),
  });
  if (!res.ok) throw new Error(`agent registration failed: ${res.status}`);
  const body = (await res.json()) as { kid: string };
  return body.kid;
}

export async function ensureAgentIdentity(args: {
  agentId: string;
  apiKey?: string;
  endpoint: string;
  identityDir?: string;
  register?: boolean;
  registerFn?: RegisterFn;
}): Promise<AgentIdentity> {
  const idDir = args.identityDir ?? defaultIdentityDir();
  const ns = namespace(args.apiKey, args.endpoint);
  const ck = `${ns}:${args.agentId}`;

  const cached = cache.get(ck);
  if (cached) return cached;

  const fromEnvIdentity = fromEnv(args.agentId);
  if (fromEnvIdentity) {
    let identity = fromEnvIdentity;
    if (args.register !== false) {
      const jwk = await publicJwk(identity.seedHex, identity.kid);
      const kid = await registerWithVerifier({
        endpoint: args.endpoint,
        apiKey: args.apiKey,
        agentId: args.agentId,
        publicJwk: jwk,
        registerFn: args.registerFn,
      });
      identity = { ...identity, kid };
    }
    cache.set(ck, identity);
    return identity;
  }

  const filePath = identityFile(idDir, ns, args.agentId);
  const existing = await loadFile(filePath);
  if (existing) {
    if (args.register !== false) {
      const jwk = await publicJwk(existing.seedHex, existing.kid);
      existing.kid = await registerWithVerifier({
        endpoint: args.endpoint,
        apiKey: args.apiKey,
        agentId: args.agentId,
        publicJwk: jwk,
        registerFn: args.registerFn,
      });
    }
    cache.set(ck, existing);
    return existing;
  }

  const { seedHex } = await generateSigner();
  let kid = newId("key");
  if (args.register !== false) {
    const jwk = await publicJwk(seedHex, kid);
    kid = await registerWithVerifier({
      endpoint: args.endpoint,
      apiKey: args.apiKey,
      agentId: args.agentId,
      publicJwk: jwk,
      registerFn: args.registerFn,
    });
  }
  const identity: AgentIdentity = { agentId: args.agentId, kid, seedHex };
  await saveFile(filePath, identity);
  cache.set(ck, identity);
  return identity;
}

/** Clear in-memory cache (tests). */
export function clearIdentityCache(): void {
  cache.clear();
}

File: sdk/python/src/tap_sdk/identity.py
"""Local agent identity storage and Verifier registration.

Developers identify agents in code (``agent_id="finance-bot-v1"``). The SDK
generates an Ed25519 keypair on first use, persists it under ``~/.tap``, and
registers the public key with the Verifier via ``POST /v1/agents/register``.
No console step is required.
"""
from __future__ import annotations

import hashlib
import json
import os
import threading
from dataclasses import dataclass
from pathlib import Path

import httpx

from typing import Callable

from .core import generate_signer, load_signer, new_id, public_jwk

RegisterFn = Callable[..., str]

_REGISTRY_LOCK = threading.Lock()
_IDENTITY_CACHE: dict[str, AgentIdentity] = {}


@dataclass(frozen=True)
class AgentIdentity:
    agent_id: str
    kid: str
    seed_hex: str


def default_identity_dir() -> Path:
    override = os.environ.get("TAP_IDENTITY_DIR")
    if override:
        return Path(override).expanduser()
    return Path.home() / ".tap" / "identities"


def _namespace(api_key: str | None, endpoint: str) -> str:
    digest = hashlib.sha256(f"{endpoint}\0{api_key or ''}".encode()).hexdigest()
    return digest[:16]


def _cache_key(namespace: str, agent_id: str) -> str:
    return f"{namespace}:{agent_id}"


def _identity_file(identity_dir: Path, namespace: str, agent_id: str) -> Path:
    safe = agent_id.replace("/", "_").replace("\\", "_")
    return identity_dir / namespace / f"{safe}.json"


def _load_file(path: Path) -> AgentIdentity | None:
    if not path.is_file():
        return None
    data = json.loads(path.read_text())
    return AgentIdentity(
        agent_id=data["agent_id"],
        kid=data["kid"],
        seed_hex=data["seed_hex"],
    )


def _save_file(path: Path, identity: AgentIdentity) -> None:
    path.parent.mkdir(parents=True, exist_ok=True)
    path.write_text(
        json.dumps(
            {
                "agent_id": identity.agent_id,
                "kid": identity.kid,
                "seed_hex": identity.seed_hex,
            },
            indent=2,
        )
        + "\n",
    )
    try:
        os.chmod(path, 0o600)
    except OSError:
        pass


def _from_env(agent_id: str) -> AgentIdentity | None:
    env_agent = os.environ.get("TAP_AGENT_ID") or os.environ.get("TAP_AGENT_ID")
    seed = os.environ.get("TAP_PRIVATE_KEY") or os.environ.get("TAP_PRIVATE_KEY_HEX")
    kid = os.environ.get("TAP_KID")
    if not seed or not kid:
        return None
    if env_agent and env_agent != agent_id:
        return None
    return AgentIdentity(agent_id=agent_id, kid=kid, seed_hex=seed.replace("0x", ""))


def register_with_verifier(
    *,
    endpoint: str,
    api_key: str | None,
    agent_id: str,
    public_jwk_dict: dict,
    timeout: float = 10.0,
    post: RegisterFn | None = None,
) -> str:
    if post is not None:
        return post(
            endpoint=endpoint,
            api_key=api_key,
            agent_id=agent_id,
            public_jwk_dict=public_jwk_dict,
            timeout=timeout,
        )
    headers: dict[str, str] = {}
    if api_key:
        headers["X-API-Key"] = api_key
    with httpx.Client(timeout=timeout) as http:
        resp = http.post(
            f"{endpoint.rstrip('/')}/v1/agents/register",
            json={"agent_id": agent_id, "public_jwk": public_jwk_dict},
            headers=headers,
        )
        resp.raise_for_status()
        return resp.json()["kid"]


def ensure_agent_identity(
    *,
    agent_id: str,
    api_key: str | None,
    endpoint: str,
    identity_dir: Path | None = None,
    register: bool = True,
    register_fn: RegisterFn | None = None,
) -> AgentIdentity:
    """Load or create a signing identity for ``agent_id`` and register it with TAPClient."""
    id_dir = identity_dir or default_identity_dir()
    ns = _namespace(api_key, endpoint)
    ck = _cache_key(ns, agent_id)

    with _REGISTRY_LOCK:
        cached = _IDENTITY_CACHE.get(ck)
        if cached:
            return cached

        from_env = _from_env(agent_id)
        if from_env:
            if register:
                sk = load_signer(from_env.seed_hex)
                kid = register_with_verifier(
                    endpoint=endpoint,
                    api_key=api_key,
                    agent_id=agent_id,
                    public_jwk_dict=public_jwk(sk, from_env.kid),
                    post=register_fn,
                )
                from_env = AgentIdentity(agent_id=agent_id, kid=kid, seed_hex=from_env.seed_hex)
            _IDENTITY_CACHE[ck] = from_env
            return from_env

        path = _identity_file(id_dir, ns, agent_id)
        existing = _load_file(path)
        if existing:
            sk = load_signer(existing.seed_hex)
            kid = existing.kid
            if register:
                kid = register_with_verifier(
                    endpoint=endpoint,
                    api_key=api_key,
                    agent_id=agent_id,
                    public_jwk_dict=public_jwk(sk, existing.kid),
                    post=register_fn,
                )
            identity = AgentIdentity(agent_id=agent_id, kid=kid, seed_hex=existing.seed_hex)
            _IDENTITY_CACHE[ck] = identity
            return identity

        sk, seed_hex = generate_signer()
        kid = new_id("key")
        if register:
            kid = register_with_verifier(
                endpoint=endpoint,
                api_key=api_key,
                agent_id=agent_id,
                public_jwk_dict=public_jwk(sk, kid),
                post=register_fn,
            )
        identity = AgentIdentity(agent_id=agent_id, kid=kid, seed_hex=seed_hex)
        _save_file(path, identity)
        _IDENTITY_CACHE[ck] = identity
        return identity
